model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/organization/InfinispanOrganizationProvider.java

```diff
@@ -0,0 +1,222 @@
+        } else if (managedOrganizations.containsKey(id)) {
+            return managedOrganizations.get(id);
+        }
+        OrganizationAdapter adapter = new OrganizationAdapter(cached, realmCache, jpaOrgDelegate);
+        managedOrganizations.put(id, adapter);
+        return adapter;
+    }
+
+    @Override
+    public OrganizationModel getByDomainName(String domainName) {
+        String cacheKey = getRealm().getId() + "+.org.domain.name." + domainName;
+        CachedOrganization cached = realmCache.getCache().get(cacheKey, CachedOrganization.class);
+        String realmId = getRealm().getId();
+        if (cached != null && !cached.getRealm().equals(realmId)) {
+            cached = null;
+        }
+
+        if (cached == null) {
+            Long loaded = realmCache.getCache().getCurrentRevision(cacheKey);
+            OrganizationModel model = jpaOrgDelegate.getByDomainName(domainName);
+            if (model == null) return null;
+            if (realmCache.getInvalidations().contains(model.getId())) return model;
+            cached = new CachedOrganization(loaded, getRealm(), model);
+            realmCache.getCache().addRevisioned(cached, realmCache.getStartupRevision());
+
+        // no need to check for realm invalidation as IdP changes are handled by events within InfinispanOrganizationProviderFactory
+        } else if (realmCache.getInvalidations().contains(cached.getId())) {
+            return getCacheDelegate(jpaOrgDelegate.getByDomainName(domainName));
+        } else if (managedOrganizations.containsKey(cached.getId())) {
+            return managedOrganizations.get(cached.getId());
+        }
+        OrganizationAdapter adapter = new OrganizationAdapter(cached, realmCache, jpaOrgDelegate);
+        managedOrganizations.put(cacheKey, adapter);
+        return adapter;
+    }
+
+    @Override
+    public Stream<OrganizationModel> getAllStream(String search, Boolean exact, Integer first, Integer max) {
+        // Return cache delegates to ensure cache invalidation during write operations
+        return getCacheDelegates(jpaOrgDelegate.getAllStream(search, exact, first, max));
+    }
+
+    @Override
+    public Stream<OrganizationModel> getAllStream(Map<String, String> attributes, Integer first, Integer max) {
+        // Return cache delegates to ensure cache invalidation during write operations
+        return getCacheDelegates(jpaOrgDelegate.getAllStream(attributes, first, max));
+    }
+
+    @Override
+    public boolean remove(OrganizationModel organization) {
+        registerOrganizationInvalidation(organization.getId());
+        return jpaOrgDelegate.remove(organization);
+    }
+
+    @Override
+    public void removeAll() {
+        //TODO: won't scale, requires a better mechanism for bulk deleting organizations within a realm
+        //this way, all organizations in the realm will be invalidated ... or should it be invalidated whole realm instead?
+        getAllStream().forEach(this::remove);
```

Comment:
I would leave it as is for now - when we are able to perf test orgs then it might the case that it is better to invalidate the realm instead.